wallet/src/account.rs

```diff
@@ -129,59 +129,110 @@ impl Account {
             return Err(WalletError::NoTransactionFound(tx_id));
         }
 
-        let account_tx_id = AccountTxId::new(self.get_acount_id(), tx_id);
+        let account_tx_id = AccountTxId::new(self.get_account_id(), tx_id);
         db_tx.del_transaction(&account_tx_id)?;
 
-        self.txs.remove(&tx_id);
-
-        // TODO remove UTXO?
+        if let Some(wallet_tx) = self.txs.remove(&tx_id) {
+            self.remove_from_utxos(db_tx, &wallet_tx)?;
+        }
 
         Ok(())
     }
 
-    // TODO fix incompatibility between borrowing mut self and the database transaction
-    #[allow(dead_code)] // TODO remove
-    fn add_to_utxos<B: storage::Backend>(
+    /// Add the transaction outputs to the UTXO set of the account
+    fn add_to_utxos<B: Backend>(
         &mut self,
         db_tx: &mut StoreTxRw<B>,
-        tx: &Transaction,
+        wallet_tx: &WalletTx,
     ) -> WalletResult<()> {
+        // Only Confirmed can be added to the UTXO set
+        if match wallet_tx.get_state() {
+            TxState::Confirmed(_) => false,
+            TxState::InMempool | TxState::Conflicted(_) | TxState::Inactive => true,
+        } {
+            return Ok(());
+        }
+
+        let tx = wallet_tx.get_tx();
+
         for (i, output) in tx.outputs().iter().enumerate() {
             // Check if this output belongs to this wallet or it is watched
             if self.is_available_for_spending(output) && self.is_mine_or_watched(output) {
                 let outpoint = OutPoint::new(tx.get_id().into(), i as u32);
                 let utxo = Utxo::new(output.clone(), false, utxo::UtxoSource::Mempool);
                 self.utxo.insert(outpoint.clone(), utxo.clone());
-                let account_utxo_id = AccountOutPointId::new(self.get_acount_id(), outpoint);
+                let account_utxo_id = AccountOutPointId::new(self.get_account_id(), outpoint);
                 db_tx.set_utxo(&account_utxo_id, utxo)?;
             }
         }
         Ok(())
     }
 
+    /// Remove transaction outputs from the UTXO set of the account
+    fn remove_from_utxos<B: Backend>(
+        &mut self,
+        db_tx: &mut StoreTxRw<B>,
+        wallet_tx: &WalletTx,
+    ) -> WalletResult<()> {
+        let tx = wallet_tx.get_tx();
+        for i in 0..tx.outputs().len() {
```

Comment:
Enumerating yourself is a bad idea as it's prone to off-by-one error. Usually the better solution is something like `tx.outputs().iter().enumerate()`, which will return a pair of the value and its index, which then you can use.